File: java/flatten-array/src/main/java/Flattener.java
import java.util.*;

class Flattener {

    List<Object> flatten(List<?> list) {
        List<Object> result = new ArrayList<>();

        for (Object item : list) {
            if (item instanceof List<?> collection) {
                result.addAll(flatten(collection));
            } else if (item != null) {
                result.add(item);
            }
        }

        return result;
    }

}

File: kotlin/flatten-array/src/main/kotlin/Flattener.kt
object Flattener {
    fun flatten(source: Collection<Any?>): List<Any> {
        return source.flatMap {
            if (it is Collection<Any?>) {
                flatten(it)
            } else {
                if (it == null) {
                    emptyList()
                } else {
                    listOf(it)
                }
            }
        }
    }
}
